### PR TITLE
store: implement non-block reading for Get and BatchGet under the large transaction protocol

### PR DESCRIPTION
**Pull request by tiancaiamao**
### What problem does this PR solve? 

Large transactions should not block the read operation.

We've handle the coprocessor reading in https://github.com/pingcap/tidb/pull/11986, and this commit continues with snapshot reading.

### What is changed and how it works?

Implement non-block read for `snapshot.Get()` and `snapshot.BatchGet()` function.

When those operations meet lock, `ResolveLock` and retry. 
The `SendReqCtx` operation would not block in the retry as we pass the `resolvedLocks`.

### Check List 

Tests 

 - Unit test

Code changes

 - Has exported function/method change

### CODE REVIEW

**Comment by tiancaiamao** [comment]
PTAL @coocood @lysu

**Comment by tiancaiamao** [comment]
Integration test fails, we also need to implement this in TiKV? @MyonKeminta @youjiali1995 

```
[2019-11-19T17:06:20.975Z] ----------------------------------------------------------------------
[2019-11-19T17:06:20.975Z] FAIL: snapshot_test.go:211: testSnapshotSuite.TestSkipLargeTxnLock
[2019-11-19T17:06:20.975Z] 
[2019-11-19T17:06:20.975Z] snapshot_test.go:232:
[2019-11-19T17:06:20.975Z]     // Commit txn, check the final commit ts is pushed.
[2019-11-19T17:06:20.975Z]     c.Assert(committer.commitKeys(bo, committer.keys), IsNil)
[2019-11-19T17:06:20.975Z] ... value *errors.fundamental = tikv aborts txn: Txn(InvalidTxnTso { start_ts: 412662674766430256, commit_ts: 0 }) ("tikv aborts txn: Txn(InvalidTxnTso { start_ts: 412662674766430256, commit_ts: 0 })")
[2019-11-19T17:06:20.975Z] 
[2019-11-19T17:06:20.975Z] [2019/11/20 01:06:17.109 +08:00] [INFO] [lock_resolver.go:540] ["resolveLock rollback"] [lock="key: []byte{0x78}, primary: []byte{0x78}, txnStartTS: 412662674766430256, ttl: 18, type: Put"]
[2019-11-19T17:06:20.975Z] [2019/11/20 01:06:17.110 +08:00] [INFO] [lock_resolver.go:540] ["resolveLock rollback"] [lock="key: []byte{0x79}, primary: []byte{0x78}, txnStartTS: 412662674766430256, ttl: 18, type: Put"]
[2019-11-19T17:06:20.975Z] 
[2019-11-19T17:06:20.975Z] ----------------------------------------------------------------------
```

File: store/mockstore/mocktikv/executor_test.go [hunk 1]
```diff
@@ -93,6 +93,12 @@ func (s *testExecutorSuite) TestResolvedLargeTxnLocks(c *C) {
 	// After that, the query should read the previous version data.
 	tk.MustQuery("select * from t").Check(testkit.Rows("1 1"))
 
+	// Cover BatchGet.
+	tk.MustQuery("select * from t where id in (1)").Check(testkit.Rows("1 1"))
+
+	// Cover PointGet.
+	tk.MustQuery("select * from t where id = 1").Check(testkit.Rows("1 1"))
```
Comment (coocood) [inline store/mockstore/mocktikv/executor_test.go]: Does this query use MaxTS?
Then we need also to cover PointGet with real TS.

Comment (tiancaiamao) [reply store/mockstore/mocktikv/executor_test.go]: Done

**Comment by coocood** [comment]
LGTM

File: store/tikv/snapshot.go [hunk 2]
```diff
@@ -285,10 +298,17 @@ func (s *tikvSnapshot) get(bo *Backoffer, k kv.Key) ([]byte, error) {
 	}
 
 	failpoint.Inject("snapshot-get-cache-fail", func(_ failpoint.Value) {
-		panic("cache miss")
+		if bo.ctx.Value("TestSnapshotCache") != nil {
```
Comment (lysu) [inline store/tikv/snapshot.go]: maybe using exp like `1*return(true)->return(false) ` like this https://github.com/pingcap/tidb/blob/2aa571b1abf329d336413fe195c88fe07c7b268c/session/session_test.go#L601 and check `failpoint.Value == true` at here is more gofail idiom~

**Comment by coocood** [comment]
@MyonKeminta PTAL

File: store/tikv/snapshot.go (hunk 2 above)
Comment (tiancaiamao) [reply store/tikv/snapshot.go]: In the concurrent test environment, that mock panic may affect other tests.
`bo.ctx.Value("TestSnapshotCache")` is used to address that problem. @lysu

Comment (coocood) [reply store/tikv/snapshot.go]: Or we can pass a TS and panic on equal TS.

Comment (tiancaiamao) [reply store/tikv/snapshot.go]: In `failpoint.Inject` it doesn't know which TS @coocood

**Pull request merged**